README.md

```diff
@@ -854,6 +855,139 @@ registration: {
+    ```
+Optional configurations :
+
+- **parseSchema:** Callback used to mold the JSON schema that comes back from Okta API.
+
+    ```javascript
+    /**
+     The callback function is passed 3 arguments:
+     @params : schema, onSuccess, onFailure
+     1) @param {Object} schema: json schema returned from the API.
+     2) @param {Function} onSuccess: success callback.
+     3) @param {Function} onFailure: failure callback. Note: accepts a errorObject that can be used to show form level or field level errors.
+    **/
+    parseSchema: function (schema, onSuccess, onFailure) {
+      // This example will add an additional field to the registration form
+        schema.profileSchema.properties.address = {
+          'type': 'string',
+          'description': 'Street Address',
+          'default': 'Enter your street address',
+          'maxLength': 255
+        };
+        schema.profileSchema.fieldOrder.push('address');
+        onSuccess(schema);
+    }
+    onFailure callback accepts a errorObject that can be used to show form level or field level errors
+    ```
+ - **preSubmit :** Callback used primarily to modify the request parameters sent to Okta API.
+
+    ```javascript
+    /**
+     The callback function is passed 3 arguments:
+     @params : postData, onSuccess, onFailure
+     1) @param {Object} postData: form data that will be posted to the registration API.
+     2) @param {Function} onSuccess: success callback.
+     3) @param {Function} onFailure: failure callback. Note: accepts a errorObject that can be used to show form level or field level errors.
+    **/
+    preSubmit: function (postData, onSuccess, onFailure) {
+      // This example will add @companyname.com to the email if user fails to add it during registration
+      return postData.username.indexOf('@acme.com') > -1 ? username : username + '@acme.com';
+    }
+    ```
+ - **postSubmit :** Callback used to primarily get control and to modify the behavior post submission to registration API .
+
+    ```javascript
+    /**
+     The callback function is passed 3 arguments:
+     @params : response, onSuccess, onFailure
+     1) @param {Object} response: response returned from the API post registration.
+     2) @param {Function} onSuccess: success callback.
+     3) @param {Function} onFailure: failure callback. Note: accepts a errorObject that can be used to show form level or field level errors.
+    **/
+    postSubmit: function (response, onSuccess, onFailure) {
+      // This example will log the server response to the browser console
+      console.log(response);
+      onSuccess(response);
+    }
+    ```
+- **onFailure and ErrorObject :** The onFailure callback accepts an error object that can be used to show a form level vs field level error on the registration form
+
+  ```javascript
+  // Example Form level error object
+    error = {
+        "errorSummary": "Custom form level error"
+    };
+  // Example field level error object:
+    error = {
```

Comment:
Instead of including the error object, I'd recommend demonstrating how to handle it.

```javascript
// Example usage
 preSubmit: function (postData, onSuccess, onFailure) {
  if (error) {
    onFailure('Failed to register due to: ', error.errorSummary)
  }
 }
```

Reply:
@jmelberg-okta the error object has to be constructed in this format and should be passed into the onFailure function as a param . However if the developer missed to pass this error object we construct a default error object.

Reply:
I see. I thought this was the response via the API. In that case, can we include the `error` object directly inside of the function? That would have helped me understand it better.

Similar to your implementation here -> [Registration_spec.js](https://github.com/okta/okta-signin-widget/blob/c99bce856bbf209961667465fcaa3233356c7145/test/unit/spec/Registration_spec.js#L498-L501)

Reply:
As per our discussion, I 'll clean this section up a little bit.

Reply:
```javascript

  // Example usage to show form level error
    preSubmit: function (postData, onSuccess, onFailure) {
      error = {
          "errorSummary": "Custom form level error"
      };
      onFailure(error);
   }

  // Example usage to show field level error 
    preSubmit: function (postData, onSuccess, onFailure) {
      var error = {
          "errorSummary": "API Error",
          "errorCauses": [
              {
                  "errorSummary": "Custom field level error",
                  "reason": "registration.error.address", 
                  "resource": "User",
                  "property": "address", //should match field name
                  "arguments": []
              }
          ]
      };
      onFailure(error);
   }

  // Example usage with no error object
    preSubmit: function (postData, onSuccess, onFailure) {
      // A Default form level error is shown if no error object is provided
      onFailure();
   }

  ```
@jmelberg-okta  let me know if this makes more sense

Reply:
Much better. Maybe we add a few headings to make it even more clear:

-----
####  Use the default error
```javascript
preSubmit: function (postData, onSuccess, onFailure) {
  // A Default form level error is shown if no error object is provided
  onFailure();
}
```

#### Use form level error
 ```javascript
preSubmit: function (postData, onSuccess, onFailure) {
  var error = {
    "errorSummary": "Custom form level error"
  };
  onFailure(error);
}
```

#### Use field level error
```javascript
  preSubmit: function (postData, onSuccess, onFailure) {
    var error = {
        "errorSummary": "API Error",
        "errorCauses": [
            {
                "errorSummary": "Custom field level error",
                "reason": "registration.error.address", 
                "resource": "User",
                "property": "address", //should match field name
                "arguments": []
            }
        ]
    };
    onFailure(error);
 }
```